Keas.Mvc/Services/HistoryService.cs

```diff
@@ -442,6 +453,132 @@ public async Task<History> WorkstationAccepted(Workstation workstation)
             await _context.SaveChangesAsync();
             return historyEntry;
         }
+
+        public async Task<History> KeyDeleted(Key key)
+        {
+            var user = await _securityService.GetUser();
+            var historyEntry = new History
+            {
+                Description = key.Name != null ? "Key (" + key.Name + ") Deleted by " + user.Name : "Key (" + key.Code + ") Deleted by " + user.Name,
```

Comment:
You might want to think about a utility method because this same code is repeated a tons of times and if we ever change the wording slightly it would be good to not have to change in so many places.

I think if you use one of the base classes like AssetBase you might be able to clean this up, say `GetDescription(asset, user, action)` or something where action is deleted in this case.  Each asset could have an override to `toString()` as well to be able to write it's own display so you can offload the work of null checking and such

Reply:
Do not override toString()

Create an abstract GetAssetName method on the asset base and implement it with nameof(class)

Reply:
sure that works